# Assignment 5 Space Cows Transportation



## Introduction
In this assignment, a colony of Aucks (super-intelligent alien bio-engineers) has landed on Earth and has created new species of farm animals! The Aucks are performing their experiments on Earth, and plan on transporting the mutant animals back to their home planet of Aurock. In this problem set, you will implement algorithms to figure out how the aliens should shuttle their experimental animals back across space.



## Getting Started 

Download Assignment5.zip from the website.

Please do not rename the provided files, change any of the provided helper functions, change function/method names, or delete provided docstrings. You will need to keep ```a5_cow_data.txt```, and ```a5_cow_data_1.txt```, in the same folder as ```a5.ipynb```.


## Problem Set Overview

The aliens have succeeded in breeding cows that jump over the moon! Now they want to take home their mutant cows. The aliens want to take all chosen cows back, but their spaceship has a weight limit and they can only travel a limitted number of trips they have to take across the universe. Somehow, the aliens have evolved and developed breeding technology to make cows with integer weights and IQs.

The data for the cows to be transported is stored in ```a5_cow_data.txt```, and another set of cows for another separate transport are in ```a5_cow_data_1.txt```. (You may use the two files to read data and test your implementation individually). All of your code for the problem solving in this assignment should go into ```a5.ipynb```--you need to expand the given notebook to include your Python code and discussion notes.  

For each problem, I provide some skeleton code for you to start your problem solving. Note that most of the code definitions are not complete unless I point out the completion of some certain function such as **greedy** for Problem 3. For each code cell that contains only incomplete code, you need to extend the code implementation.  

You also need to solve the problems in the order presented in this document.  That is, you should complete problem 1 first before you approach problems 2 and 3.  The solutions of the later problems are dependent on the solutions to earlier problems.  For example, in Problem 2, you need to parse a data file to create Cow objects.  The class definition of Cow for Problem 1 is needed to create the Cow objects for Problem 2.  

# Problem 1: Defining Cow Class

First we need to define a **Cow** class.  Each cow object state is described using name as a string and weight as an int. (You may check the Food class definition in the lecture notes as a reference when you are working on defining the Cow class.) 

Note that I provided some skeleton code below so that you may expand based on what is provided.  

In [4]:
class Cow(object):
    
    def __init__(self, name, weight, IQ):
        self.name = name
        self.weight = weight
        self.IQ = IQ
        self.transported = 0  # Optional attribute to track if cow has been transported

    def __str__(self):
        return f"Cow {self.name}: Weight {self.weight}, IQ {self.IQ}"

    def getName(self):
        return self.name

    def getWeight(self):
        return self.weight

    def getIQ(self):
        return self.IQ

    def getDensity(self):
        # Calculates and returns IQ per unit weight as a measure of "density"
        return self.IQ / self.weight if self.weight != 0 else 0

# Test code
mary = Cow('mary', 3, 120)
print(mary)              # Expected: "Cow mary: Weight 3, IQ 120"
print(mary.getDensity()) # Expected: 40.0
print("\n")
# Additional test code to check all methods
print(mary.getName())    # Expected: 'mary'
print(mary.getWeight())  # Expected: 3
print(mary.getIQ())      # Expected: 120

Cow mary: Weight 3, IQ 120
40.0


mary
3
120


Note the test code provided above doesn't provide a throughout checking and evaluation of your class definition.  It only checks the __init__, __str__, and getDensity methods.  You should extend the test code to evaluate other methods you define in the class definition.   The output of the above test code could be like

<code>
Cow mary: Weight 3, IQ 120
40.0
</code>

The first output is determined by the definition of __str__ in the class definition. 

# Problem 2: Loading Cow Data

Second we need to load the cow data from the data file ```a5_cow_data.txt```.
The file ```a5_cow_data_1.txt``` is given as another file that you can read and test from, but for now, just work with a1_cow_data.txt.

You can expect the data to be formatted in triples of ```x,y,z``` on each line, where ```x``` is the name of the cow, ```y``` is a number indicating how much the cow weighs in tons, and ```z``` is a number indicating the cow's IQ value. Here are the few lines
of ```a5_cow_data.txt```: 

<code>
Maggie,3,165
Herman,7,126
Betsy,9,122
Oreo,6,104
Moo Moo,3,151
Milkshake,2,117
Millie,5,84
Lola,2,131
Florence,2,101
Henrietta,9,106
</code>

You can assume that all the cows have unique names.
Hint: If you don’t remember how to read lines from a file, check out the online python documentation, which has a chapter on **Input and Output** that includes file I/O here: https://docs.python.org/3/tutorial/inputoutput.html

Some functions that may be helpful:

<code>
str.split
open
file.readline
file.close
</code>



In [9]:
! cat "a5_cow_data.txt"

Maggie,3,165
Herman,7,126
Betsy,9,122
Oreo,6,104
Moo Moo,3,151
Milkshake,2,117
Millie,5,84
Lola,2,131
Florence,2,101
Henrietta,9,106


In [10]:
! cat "a5_cow_data_1.txt"


Miss Moo-dy,3,172
Milkshake,4,102
Lotus,10,149
Miss Bella,2,103
Horns,9,81
Betsy,5,97
Rose,3,155
Dottie,6,91


In [6]:
def load_cows(filename):
    # First create a list to store Cow objects
    data = []
    
    # To open and read the file
    with open(filename, 'r') as f:
        # Read each line in the file
        for line in f:
            # To remove any surrounding whitespace or newline characters
            line = line.strip()
            
            # Split the line by commas to extract name, weight, and IQ
            if line:
                name, weight, iq = line.split(',')
                # Create a Cow object with parsed values
                cow = Cow(name, int(weight), int(iq))
                # Add the Cow object to the data list
                data.append(cow)
    return data

# Test code
data = load_cows("a5_cow_data.txt")
for cow in data:
    print(cow)


Cow Maggie: Weight 3, IQ 165
Cow Herman: Weight 7, IQ 126
Cow Betsy: Weight 9, IQ 122
Cow Oreo: Weight 6, IQ 104
Cow Moo Moo: Weight 3, IQ 151
Cow Milkshake: Weight 2, IQ 117
Cow Millie: Weight 5, IQ 84
Cow Lola: Weight 2, IQ 131
Cow Florence: Weight 2, IQ 101
Cow Henrietta: Weight 9, IQ 106


By running the above test code, your code should get output like below
<code>
 Cow Maggie: Weight 3, IQ 165
 Cow Herman: Weight 7, IQ 126
 Cow Betsy: Weight 9, IQ 122
 Cow Oreo: Weight 6, IQ 104
 Cow Moo Moo: Weight 3, IQ 151
 Cow Milkshake: Weight 2, IQ 117
 Cow Millie: Weight 5, IQ 84
 Cow Lola: Weight 2, IQ 131
 Cow Florence: Weight 2, IQ 101
 Cow Henrietta: Weight 9, IQ 106
</code>

 
Again, I would like to remind you that the __str__ method of the Cow object represented by data[i] is called when print(data[i]) is invoked. 


### Problem 3: Greedy Cow Transport 

One way of transporting cows is to always pick the cow that has the most intelligence density (IQ/weight) onto the spaceship first. This is an example of a ```greedy``` algorithm.  You may choose a criterion to use, which you think suitable to accomplish the goal ---to transport the maximum intelligence values of cows back home. 

Implement a greedy algorithm for transporting the cows back across space in
**greedy_cow_transport**. The constraints include the weight limit for each space trip and the total number of trips the aliens can make.  The result should be a triple composed of three values: the first value presents the total sum of the IQs of the cows transported, the second value presents the total sum of the weight values of the transported cows, and the third value presents a list of lists, with each inner list containing cows (cow objects) transported on a particular trip.



To facilitate your problem solving, I provide a function definition of greedy, which is complete. (It means that you do NOT need to change anything.  I also provide some skeleton code for **greedy_cow_transport** including the function call of greedy on line 35. The function definition of **greedy** is based on the greedy algorithm we studied in the lecture.  



In [8]:

#Implementation of Flexible Greedy 
def greedy(cows, maxCost, keyFunction):
    """
    Uses a greedy approach based on a criterion to 
    determine a list of Cow objects 
    to take on a single trip by a spaceship that can carry 
    a certain amount of weight.
    
    Parameters:
        cows - a list of Cow objects
        maxCost - should be a positive int to indicate the maximum weight (tons) the trip can do
        keyFunction - should be a function that is used to sort the cows
                        and it maps an item to a number
    Returns:
        result - a list of Cows chosen to be transported by a trip
        totalValue - an int value to keep track of the sum of IQ values of the transported Cow objects
        totalCost - an int value to keep track of the sum of weights of the transfored Cow objects
    """
    
    #Attention check sorted function documentation
    itemsCopy = sorted(cows, key = keyFunction,  
                       reverse = True)
    result = []
    
    totalValue, totalCost = 0, 0
    
    for i in range(len(itemsCopy)): #Attention
        if (totalCost+itemsCopy[i].getWeight()) <= maxCost:  #Attention 
            result.append(itemsCopy[i])
            totalCost += itemsCopy[i].getWeight()
            totalValue += itemsCopy[i].getIQ()
            
    return (result, totalValue, totalCost)


In [10]:
def greedy_cow_transport(cows, oneTripWeightLimit=10, numberOfTrips=3):
    trips = []
    totalValue = 0
    totalCost = 0
    
    remaining_cows = cows[:]
    
    while numberOfTrips > 0 and remaining_cows:
        # We can use the greedy function to get the best cows for this trip based on their IQ density
        result, oneTripValue, oneTripCost = greedy(remaining_cows, oneTripWeightLimit, Cow.getDensity)
        
        # However, if no cows can be transported in this trip, then the loop breaks
        if not result:
            break
        
        # Add the trip's cows to trips
        trips.append(result)
        
        # Update totalValue and totalCost
        totalValue += oneTripValue
        totalCost += oneTripCost
        
        # Remove transported cows from remaining_cows
        remaining_cows = [cow for cow in remaining_cows if cow not in result]
        
        # Decrement the number of trips left
        numberOfTrips -= 1

    return totalValue, totalCost, trips

# Test code
cows = load_cows("a5_cow_data.txt")
totalValue, totalCost, trips = greedy_cow_transport(cows)

for i in range(len(trips)):
    print("Trip " + str(i) + ":")
    for cow in trips[i]:
        print(cow)
    print("\n")

print("Total IQs transported =", totalValue)
print("Total weights (tons) transported =", totalCost)


Trip 0:
Cow Lola: Weight 2, IQ 131
Cow Milkshake: Weight 2, IQ 117
Cow Maggie: Weight 3, IQ 165
Cow Florence: Weight 2, IQ 101


Trip 1:
Cow Moo Moo: Weight 3, IQ 151
Cow Herman: Weight 7, IQ 126


Trip 2:
Cow Oreo: Weight 6, IQ 104


Total IQs transported = 895
Total weights (tons) transported = 25


The output of the above test code should be like below

<code>
Trip 0:
 Cow Lola: Weight 2, IQ 131
 Cow Milkshake: Weight 2, IQ 117
 Cow Maggie: Weight 3, IQ 165
 Cow Florence: Weight 2, IQ 101


Trip 1:
 Cow Moo Moo: Weight 3, IQ 151
 Cow Herman: Weight 7, IQ 126


Trip 2:
 Cow Oreo: Weight 6, IQ 104


Total IQs transported = 895
Total weights (tons) transported = 25
</code>


### Problem 4: Writeup

Answer the following questions:

<li>Does the greedy algorithm return the optimal solution? Think: How do you evaluate greedy algorithms in general?  How about this case? </li>
<li>If yes, why?
<li>If not, what could be a solution that can find the optimal solution?  </li>

Note that you can write your answers to the questions in this notebook document with your code implementations.   

1. **Does the greedy algorithm return the optimal solution?**
With thorough analysis, I would say that the greedy algorithm doesn’t ensure the optimal solution here. While it selects cows based on the highest IQ density, it does not account for the overall combination of cows across trips, which may lead to a suboptimal total IQ. Essentially, the strategy might miss the big picture of what’s best for the entire journey.

2. **How do you evaluate greedy algorithms in general? How about this case?**
Greedy algorithms are generally valued for being fast and delivering workable solutions. Here, the greedy approach is quick and simple, making it practical for time-sensitive tasks. However, since it prioritizes one metric (IQ density), it doesn’t always lead to the highest possible IQ total, as it might ignore combinations that would fit better across multiple trips the Aucks would make and that the approximation quality may vary since it might not yield the maximum possible IQ sum due to the limitations of density-based selection

3. **If not, what could be a solution that can find the optimal solution?**
For the absolute best solution, a dynamic programming algorithm would be more reliable. Dynamic programming (DP) is suitable for the cow transport problem because it allows us to explore and record optimal solutions for subproblems—such as finding the best combination of cows for each trip under the weight constraint. DP efficiently finds the solution by breaking the problem down into overlapping subproblems and storing solutions for these subproblems to avoid redundant calculations.

Here is the new solution for problem 3 with the dynamic programming:

In [12]:
from itertools import combinations

def dp_cow_transport(cows, oneTripWeightLimit=10, numberOfTrips=3):
    """
    Uses a dynamic programming approach to determine the best allocation of cows 
    to maximize the total IQ while respecting the constraints of trip weight 
    and trip count.
    
    Parameters:
    cows - a list of Cow objects
    oneTripWeightLimit - the maximum weight the spaceship can carry in one trip
    numberOfTrips - the total number of trips allowed
    
    Returns:
    A tuple with the total IQ of transported cows, the total weight, and a list of 
    lists representing each trip.
    """
    # Step 1: Prepare a dictionary to store optimal results for each weight and cow combination
    dp_table = {}

    def best_trip(cows_list, weight_limit):
        """
        Finds the optimal subset of cows within a single trip under weight constraint.
        """
        # Check if result is already computed
        if (tuple(cows_list), weight_limit) in dp_table:
            return dp_table[(tuple(cows_list), weight_limit)]

        max_iq = 0
        best_subset = []

        # Iterate through all subsets of cows to find the best trip configuration
        for r in range(1, len(cows_list) + 1):
            for subset in combinations(cows_list, r):
                total_weight = sum(cow.getWeight() for cow in subset)
                if total_weight <= weight_limit:
                    total_iq = sum(cow.getIQ() for cow in subset)
                    if total_iq > max_iq:
                        max_iq = total_iq
                        best_subset = list(subset)
        
        dp_table[(tuple(cows_list), weight_limit)] = (best_subset, max_iq, sum(cow.getWeight() for cow in best_subset))
        return best_subset, max_iq, sum(cow.getWeight() for cow in best_subset)

    trips = []
    total_iq = 0
    total_weight = 0

    remaining_cows = cows[:]
    while numberOfTrips > 0 and remaining_cows:
        # Find the best possible trip for the remaining cows
        best_trip_cows, trip_iq, trip_weight = best_trip(remaining_cows, oneTripWeightLimit)
        
        # Update trips and remaining cows
        trips.append(best_trip_cows)
        total_iq += trip_iq
        total_weight += trip_weight
        
        # Remove transported cows from remaining cows list
        remaining_cows = [cow for cow in remaining_cows if cow not in best_trip_cows]
        
        numberOfTrips -= 1  # Decrement the trip counter

    return total_iq, total_weight, trips

# Testing the DP-based function
cows = load_cows("a5_cow_data.txt")
total_iq, total_weight, trips = dp_cow_transport(cows)
for i, trip in enumerate(trips):
    print(f"Trip {i + 1}:")
    for cow in trip:
        print(cow)
    print("\n")
print("Total IQs transported =", total_iq)
print("Total weights (tons) transported =", total_weight)

Trip 1:
Cow Maggie: Weight 3, IQ 165
Cow Moo Moo: Weight 3, IQ 151
Cow Milkshake: Weight 2, IQ 117
Cow Lola: Weight 2, IQ 131


Trip 2:
Cow Herman: Weight 7, IQ 126
Cow Florence: Weight 2, IQ 101


Trip 3:
Cow Betsy: Weight 9, IQ 122


Total IQs transported = 913
Total weights (tons) transported = 28


# Turn-in
You need to turn in below for your submission:

* Your notebook file that contains the code and presentation.  Note that you need to execute the code cells to generate output that should be similar to the output examples presented in this document.  My running environment is different from yours.  To make sure I evaluate your notebook fairly, you should provide me the output you ran at your side. After you run your code cells, you can save the notebook file.   
* Any other supplementary documents you want to submit to D2L Assignments folder 

You need to package the files into a zip archive and upload the zip file to D2L assignment folder <b>Assignment 5</b>